pkg/kv/kvtest/store.go

```diff
@@ -0,0 +1,269 @@
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("previous_value", func(t *testing.T) {
+		// setup initial value
+		key := []byte("setIfKey1")
+		value := []byte("val1")
+		err := store.Set(ctx, key, value)
+		if err != nil {
+			t.Fatalf("failed to set key=%s, value=%s: %s", key, value, err)
+		}
+
+		// should set new value if old value is kept
+		newValue := []byte("val2")
+		err = store.SetIf(ctx, key, newValue, value)
+		if err != nil {
+			t.Fatalf("failed to setIf key=%s value=%s pred=%s: %s", key, newValue, value, err)
+		}
+
+		// should NOT set new value if old value is not kept
+		newValue = []byte("val3")
+		err = store.SetIf(ctx, key, newValue, value)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("setIf err=%s, key=%s, value=%s, pred=%s, expected err=%s", err, key, newValue, value, kv.ErrNotFound)
+		}
+
+		// should NOT set new value if try conditional nil
+		newValue = []byte("val4")
+		err = store.SetIf(ctx, key, newValue, nil)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("setIf err=%s, key=%s, value=%s, pred=%s, expected err=%s", err, key, newValue, value, kv.ErrNotFound)
+		}
+	})
+
+	t.Run("no_previous_key", func(t *testing.T) {
+		// setIf without previous value
+		key := []byte("setIfKey2")
+		value := []byte("val1")
+		err := store.SetIf(ctx, key, value, nil)
+		if err != nil {
+			t.Fatalf("setIf without previous key - key=%s, value=%s: %s", key, value, err)
+		}
+
+		// should fail if we try to do the same again - as the value updated to 'val1'
+		err = store.SetIf(ctx, key, value, nil)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("setIf with nil while key/value exists - err=%v, key=%s value=%s, expected err=%s",
+				err, key, value, kv.ErrNotFound)
+		}
+	})
+}
+
+func testStoreScan(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// setup sample data
+	const samplePrefix = "scan"
+	const sampleItems = 10
+	sampleData := setupSampleData(t, ctx, store, samplePrefix, sampleItems)
+
+	t.Run("full", func(t *testing.T) {
+		scanKeyPrefix := []byte("scan-key")
+		scan, err := store.Scan(ctx, scanKeyPrefix)
+		if err != nil {
+			t.Fatal("failed to scan", err)
+		}
+		defer scan.Close()
+
+		var entries []kv.Entry
+		for scan.Next() {
+			ent := scan.Entry()
+			switch {
+			case ent == nil:
+				t.Fatal("scan got nil entry")
+			case ent.Key == nil:
+				t.Fatal("Key is nil while scan item", len(entries))
+			case ent.Value == nil:
+				t.Fatal("Value is nil while scan item", len(entries))
+			}
+			if !bytes.HasPrefix(ent.Key, scanKeyPrefix) {
+				break
+			}
+			entries = append(entries, *ent)
+		}
+		if err := scan.Err(); err != nil {
+			t.Fatal("scan ended with an error", err)
+		}
+		if diff := deep.Equal(entries, sampleData); diff != nil {
+			t.Fatal("scan data didn't match:", diff)
+		}
+	})
+
+	t.Run("part", func(t *testing.T) {
+		scan, err := store.Scan(ctx, []byte("scan-key-5"))
```

Comment:
This test only works because
```go
 	const sampleItems = 10
```
above.  It would fail with
```go
 	const sampleItems = 100
```

OTOH it would be _much_ better, because it would scan multiple items.